src/main/java/org/apache/commons/collections4/iterators/IteratorChain.java

```diff
@@ -183,9 +188,12 @@ private void checkLocked() {
     @Override
     public boolean hasNext() {
         lockChain();
-        updateCurrentIterator();
-        lastUsedIterator = currentIterator;
-        return currentIterator.hasNext();
+        if (cachedHasNextValue == null) {
+            updateCurrentIterator();
+            lastUsedIterator = currentIterator;
+            cachedHasNextValue = currentIterator.hasNext();
```

Comment:
This still results in two `hasNext()` calls for each element: the first happens inside `updateCurrentIterator()`, and the second is the explicit call here. That means the double-invocation issue (and its performance impact) remains. We should refactor so that `hasNext()` is only called **once** per element.